temporal-test-server/src/main/java/io/temporal/internal/testservice/TestWorkflowMutableStateImpl.java

```diff
@@ -2074,48 +2090,68 @@ private void timeoutActivity(long scheduledEventId, TimeoutType timeoutType, int
   }
 
   @Override
-  public void startNexusTask(long scheduledEventId, RespondNexusTaskCompletedRequest request) {
+  public void startNexusOperation(
+      long scheduledEventId, String clientIdentity, StartOperationResponse.Async resp) {
     update(
         ctx -> {
           StateMachine<NexusOperationData> operation = getPendingNexusOperation(scheduledEventId);
-          operation.action(StateMachines.Action.START, ctx, request, 0);
-          operation.getData().identity = request.getIdentity();
+          operation.action(StateMachines.Action.START, ctx, resp, 0);
+          operation.getData().identity = clientIdentity;
+          scheduleWorkflowTask(ctx);
         });
   }
 
   @Override
-  public void completeNexusTask(long scheduledEventId, RespondNexusTaskCompletedRequest request) {
+  public void cancelNexusOperation(NexusOperationRef ref) {
     update(
         ctx -> {
-          StateMachine<NexusOperationData> operation = getPendingNexusOperation(scheduledEventId);
-          throwIfOperationTokenDoesntMatch(request.getTaskToken(), operation.getData());
-          if (request.getResponse().hasCancelOperation()) {
-            operation.action(Action.CANCEL, ctx, request, 0);
-          } else {
-            operation.action(StateMachines.Action.COMPLETE, ctx, request, 0);
+          StateMachine<NexusOperationData> operation =
+              getPendingNexusOperation(ref.getScheduledEventId());
+          if (!operationRefMatches(ref, operation.getData())
+              || !operationInFlight(operation.getState())) {
+            return;
           }
-          // nexusOperations.remove(scheduledEventId); // TODO(pj): server currently does not delete
+          operation.action(Action.CANCEL, ctx, null, 0);
+          nexusOperations.remove(ref.getScheduledEventId());
           scheduleWorkflowTask(ctx);
-          ctx.unlockTimer("completeNexusTask");
+          ctx.unlockTimer("cancelNexusOperation");
         });
   }
 
   @Override
-  public void failNexusTask(long scheduledEventId, RespondNexusTaskFailedRequest request) {
+  public void completeNexusOperation(NexusOperationRef ref, Payload result) {
     update(
         ctx -> {
-          StateMachine<NexusOperationData> operation = getPendingNexusOperation(scheduledEventId);
-          throwIfOperationTokenDoesntMatch(request.getTaskToken(), operation.getData());
-          operation.action(StateMachines.Action.FAIL, ctx, request, 0);
+          StateMachine<NexusOperationData> operation =
+              getPendingNexusOperation(ref.getScheduledEventId());
+          if (!operationRefMatches(ref, operation.getData())) {
```

Comment:
When completing an operation asynchronously, the completion token shouldn't include an attempt or cancelation info. It's just a reference to the operation state machine (namespace, workflow ID, run ID, scheduled event ID).
I would consider splitting the concept of a task token and an operation reference because those have different semantics.
A task token is valid for a single attempt, while the reference is valid for the lifetime of the operation.